### PR TITLE
[OpInfo] Added ReductionOpInfo subclass of OpInfo and ported sum test

### PR DESCRIPTION
**Pull request by heitorschueroff**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #63978
* #62900
* #62899
* __->__ #62737

ReductionOpInfo is a specialization of OpInfo for reduction operators. For now, it is designed to work with reductions that return a single tensor and that reduce all elements along one or more dimensions to a single value. In particular this excludes operators such as `max` and `min` that return multiple tensors and `quantile` that can return multiple values.

fixes https://github.com/pytorch/pytorch/issues/49746

Differential Revision: [D30406568](https://our.internmc.facebook.com/intern/diff/D30406568)

### CODE REVIEW

File: torch/testing/_internal/common_methods_invocations.py [hunk 1]
```diff
@@ -42,6 +42,15 @@
     import scipy.special
 
 
+# Reasonable testing sizes for dimensions
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:45]: Moving these here is a good idea

File: torch/testing/_internal/common_methods_invocations.py [hunk 2]
```diff
@@ -750,9 +758,165 @@ def default_test_dtypes(self, device_type):
...
+    yield make_tensor((2,), device, dtype, requires_grad=requires_grad)
+    yield make_tensor((2, 3), device, dtype, requires_grad=requires_grad, noncontiguous=True)
+    yield make_tensor((3, 2, 1, 2, 2), device, dtype, requires_grad=requires_grad)
+
+
+# Generates a subset of possible dim and keepdim kwargs for a tensor
+# with ndim dims appropriate for testing. If supports_multiple_dims
+# is True (default) then dim kwarg can be a list of dims.
+def _generate_reduction_kwargs(ndim, supports_multiple_dims=True):
+    for keepdim in [True, False]:
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:773]: Should this also yield a no kwargs case?

Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:773]: OK it's added explicitly below. That seems fine

File: torch/testing/_internal/common_methods_invocations.py [hunk 3]
```diff
@@ -750,9 +758,165 @@ def default_test_dtypes(self, device_type):
...
+            if ndim > 1:
+                yield {'dim': (0, ndim - 1), 'keepdim': keepdim}
+
+            # Test reducing every other dimension starting with the second
+            if ndim > 3:
+                yield {'dim': tuple(range(1, ndim, 2)), 'keepdim': keepdim}
+
+
+# Wraps sample_inputs_reduction function to provide the additional supports_multiple_dims args
+def sample_inputs_reduction_wrapper(supports_multiple_dims):
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:796]: A possibly simpler option would be to:

```
def sample_inputs_reductions(op_info, device, dtype, requires_grad, *, supports_multiple_dims, **kwargs):
```

And then the sample inputs for an OpInfo would partial the `supports_multiple_dims` kwarg like the div OpInfos partial their sample input func.

File: torch/testing/_internal/common_methods_invocations.py [hunk 4]
```diff
@@ -750,9 +758,165 @@ def default_test_dtypes(self, device_type):
...
+    # Generates sample inputs for reduction ops that contain the input tensor
+    # and dim and keepdim kwargs. If a reduction op needs to test additional
+    # args/kwargs then create a separate sample_inputs function
+    def sample_inputs_func(op_info, device, dtype, requires_grad):
+        inputs = []
+
+        for t in _generate_reduction_inputs(device, dtype, requires_grad):
+            # Add case without dim and keepdim kwargs
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:804]: This special case not needed if _generate_reduction_kwargs returned `{}`, too

File: torch/testing/_internal/common_methods_invocations.py [hunk 5]
```diff
@@ -750,9 +758,165 @@ def default_test_dtypes(self, device_type):
...
+                inputs.append(SampleInput(t, kwargs=kwargs))
+
+        return inputs
+
+    return sample_inputs_func
+
+
+# NOTE [Reductions]:
+#
+# For the purpose of testing, we relax the definition of a reduction operator
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:816]: Great comment. We can also revisit this decision later since all this is internal

File: torch/testing/_internal/common_methods_invocations.py [hunk 6]
```diff
@@ -750,9 +758,165 @@ def default_test_dtypes(self, device_type):
...
+        # generates tuples of args and kwargs to use when calling the operator
+        # and reference implementation with the given dim kwarg. Note that some
+        # tests will only use the first generated args and kwargs.
+        generate_args_kwargs: Callable = lambda t, dim=None: (yield tuple(), {}),
+
+        # Options from the OpInfo base class
+        **kwargs,
+    ):
+        # These are mutually exclusive options
+        assert not (promotes_int_to_float and promotes_int_to_int64)
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:886]: Nice checks

File: torch/testing/_internal/common_methods_invocations.py [hunk 7]
```diff
@@ -750,9 +758,165 @@ def default_test_dtypes(self, device_type):
...
+        assert not (result_dtype and promotes_int_to_float)
+        assert not (result_dtype and promotes_int_to_int64)
+
+        assert nan_policy in ('propagate', 'omit')
+
+        # ReductionOpInfo tests generate their own input tensors and the dim
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:892]: Nice comment

File: torch/testing/_internal/common_methods_invocations.py [hunk 8]
```diff
@@ -750,9 +758,165 @@ def default_test_dtypes(self, device_type):
...
+        # and keepdim parameters. Operators can provide a `generate_args_kwargs`
+        # to augment the arguments to the operator and the reference implementation
+        # or specify `sample_inputs_func` to provide their own sample inputs for
+        # the OpInfo base class. However, the test_reductions.py file will still
+        # generate its own inputs.
+        def sample_inputs_func(*args, **kwargs):
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:898]: I understand the comment but I'm not sure I understand why the structure of this function is necessary -- should the default sample inputs func just be a partial of one of the above funcs? Why the kwarg surgery here?

File: torch/testing/_internal/common_methods_invocations.py [hunk 9]
```diff
@@ -750,9 +758,165 @@ def default_test_dtypes(self, device_type):
...
+            f = sample_inputs_reduction_wrapper(supports_multiple_dims)
+            for sample in f(*args, **kwargs):
+                dim = sample.kwargs.get('dim', None)
+                for args, kwargs in generate_args_kwargs(sample.input, dim=dim):
+                    kwargs.update(sample.kwargs)
+                    result.append(SampleInput(sample.input, args=args, kwargs=kwargs))
+            return result
+
+        # Override OpInfo defaults and call base class __init__
+        kwargs.setdefault('inplace_variant', None)
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:909]: Neat

File: torch/testing/_internal/common_methods_invocations.py [hunk 10]
```diff
@@ -8350,6 +8459,20 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
     ),
 ]
 
+# Reduction operator database
+reduction_op_db: List[ReductionOpInfo] = [
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:8463]: Add it to the main ops db but then create a filtered list like unary_ufuncs does for consistency

**Review by mruberry — COMMENTED** [review]
Overall looks very good, @heitorschueroff; just a few inline questions/suggestions for your review.

File: test/test_reductions.py [hunk 11]
```diff
@@ -55,6 +57,21 @@ def _rand_shape(dim, min_size, max_size):
...
 
+    ###########################################################################
+    # ReductionOpInfo unit tests
+    ###########################################################################
+
+    @ops(reduction_op_db, allowed_dtypes=[torch.float])
+    def test_dim_default(self, device, dtype, op: ReductionOpInfo):
+        """Tests that the default behavior is to reduce all dimensions."""
+        t = make_tensor((2, 3), device, dtype)
+        args, kwargs = next(op.generate_args_kwargs(t))
```
[the remark this answers is not in the document]
Comment (heitorschueroff) [reply test/test_reductions.py:68]: the test generates dim and keepdim kwargs and passed them along with the extra args and kwargs generated by the op. In this case, we do not pass dim and keepdim to op to test the default behavior. However, I'm not checking that `generate_args_kwargs` doesn't actually generate dim and keepdim and if it did then this test would probably fail.

File: torch/testing/_internal/common_methods_invocations.py (hunk 3 above)
Comment (heitorschueroff) [reply torch/testing/_internal/common_methods_invocations.py:796]: I didn't know about `partial` but that is a great suggestion, thank you!

File: torch/testing/_internal/common_methods_invocations.py (hunk 8 above)
Comment (heitorschueroff) [reply torch/testing/_internal/common_methods_invocations.py:898]: This is necessary to augment the sample inputs from `sample_inputs_reduction` with the args and kwargs from `generate_args_kwargs`.

File: test/test_reductions.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (mruberry) [reply test/test_reductions.py:68]: Right, but isn't that worth addressing?

File: torch/testing/_internal/common_methods_invocations.py [hunk 12]
```diff
@@ -769,9 +777,163 @@ def default_test_dtypes(self, device_type):
...
+def _generate_reduction_kwargs(ndim, supports_multiple_dims=True):
+    """Generates a subset of all valid dim and keepdim kwargs given ndim that
+    is appropriate for testing reduction operators.
+    """
+
+    # Test default dim and keepdim
+    yield {}
+
+    # Test reducing inner and outer most dimensions
+    yield {'dim': 0, 'keepdim': True}
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:797]: Like if this were moved above the first yield wouldn't the test above fail?

File: torch/testing/_internal/common_methods_invocations.py (hunk 8 above)
Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:915]: I guess I can buy that. An alternative design might have had `sample_inputs_reduction` accept the `generate_args_kwargs` function as an input and perform this logic itself

File: torch/testing/_internal/common_methods_invocations.py [hunk 13]
```diff
@@ -8350,6 +8459,20 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
     ),
 ]
 
+# Reduction operator database
+reduction_op_db: List[ReductionOpInfo] = [
+    ReductionOpInfo(
```
[the remark this answers is not in the document]
Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:8736]: Add this alphabetically (as best you can since the list isn't alphabetical) instead of at the end of the sequence to help avoid/prevent merge conflicts

File: test/test_reductions.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (mruberry) [reply test/test_reductions.py:68]: OK my mistake I was conflating two functions (there are a lot of functions). I guess it's reasonable to assume `generate_args_kwargs` won't generate `dim` or `keepdim`.

**Review by mruberry — APPROVED** [review]
A few small suggestions inline for you to consider, but overall things look good. Hurray ReductionInfos!

File: torch/testing/_internal/common_methods_invocations.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (heitorschueroff) [reply torch/testing/_internal/common_methods_invocations.py:8736]: That was my goals with putting it into a reductions_op_db initially, but I guess we should just do a refactor of common_methods_invocations at some point. I'll add them alphabetically.

File: torch/testing/_internal/common_methods_invocations.py (hunk 8 above)
Comment (heitorschueroff) [reply torch/testing/_internal/common_methods_invocations.py:915]: Yes, that need actually came up later when porting other ops so I might do that. Thanks for the suggestion.

**Comment by heitorschueroff** [comment]
@heitorschueroff has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30406568).

**Comment by heitorschueroff** [comment]
@heitorschueroff has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30406568).

**Comment by heitorschueroff** [comment]
@heitorschueroff has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30406568).

**Comment by heitorschueroff** [comment]
@heitorschueroff has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30406568).

**Comment by heitorschueroff** [comment]
@heitorschueroff has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30406568).
